### PR TITLE
Introduce CompositeOnnxModel

### PR DESCRIPTION
**Pull request by devang-ml**
Introduce CompositeOnnxModel to represent models with encoder and decoder subcomponents as individual OnnxModels.

Separate PRs will introduce support for creating CompositeOnnxModel and converting CompositeOnnxModel into a OnnxModel

### CODE REVIEW

File: olive/model.py [hunk 1]
```diff
@@ -77,6 +78,12 @@ def prepare_session(
         """
         raise NotImplementedError()
 
+    def set_compsite_parent(self, cp):
```
Comment (jambayk) [inline olive/model.py:81]: nit: typo

File: olive/passes/olive_pass.py [hunk 2]
```diff
@@ -336,6 +336,11 @@ def run(self, model: OliveModel, output_model_path: str, point: Optional[Dict[st
             return DistributedOnnxModel(
                 output_filepaths, model.name, version=model.version, inference_settings=model.inference_settings
             )
+        elif isinstance(model, CompositeOnnxModel):
+            components = []
+            for cidx, child in enumerate(model.get_model_components()):
+                components.append(self._run_for_config(child, config, str(f"{output_model_path}_{cidx}")))
```
Comment (jambayk) [inline olive/passes/olive_pass.py:342]: Not related to this PR but looks like the distributed onnx model implementation missed appending the rank to the `output_model_path`. It is just overwriting the same path with the subsequent ranks.

Comment (devang-ml) [reply olive/passes/olive_pass.py:342]: Yes, good catch. We can address that separately.

File: olive/model.py (hunk 1 above)
Comment (devang-ml) [reply olive/model.py]: yup. Fixed.

**Review by jambayk — DISMISSED** [review]

**Review by jambayk — DISMISSED** [review]

**Review by jambayk — DISMISSED** [review]

**Comment by trajepl** [comment]
nit: please also add documents to introduce this new changes, separated PR is ok I think.

**Review by trajepl — APPROVED** [review]
Merge and resolve conflicts.

**Review by jambayk — APPROVED** [review]

**Pull request merged**